solidity/contracts/AdaptiveStakingPolicy.sol

```diff
@@ -0,0 +1,73 @@
+pragma solidity 0.5.17;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "./libraries/grant/UnlockingSchedule.sol";
+import "./GrantStakingPolicy.sol";
+import "./TokenStaking.sol";
+
+/// @title AdaptiveStakingPolicy
+/// @dev A staking policy which allows the grantee
+/// to always stake the defined minimum stake,
+/// or the unlocked amount if greater.
+///
+/// This is necessary for staking revocable token grants safely.
+/// If the entire revocable grant can be staked,
+/// the yet-to-be-unlocked amount becomes ineffective as collateral
+/// if the grant is revoked.
+/// To avoid this issue,
+/// only the unlocked amount can be staked.
+///
+/// However, grants that feature a cliff pose a problem
+/// as no tokens are unlocked until the cliff is reached.
+/// Small grants may also take a long time
+/// to unlock enough tokens to be able to stake.
+/// To permit all grants to stake from the beginning,
+/// the policy defines a minimum which can always be staked
+/// even if the grant doesn't have enough unlocked tokens.
+contract AdaptiveStakingPolicy is GrantStakingPolicy {
+    using SafeMath for uint256;
+    using UnlockingSchedule for uint256;
+    uint256 minimumStake;
+    uint256 stakeaheadTime;
+    bool useCliff;
+
+    constructor(
+        address _stakingContract,
+        uint256 minimumMultiplier,
+        uint256 _stakeaheadTime,
+        bool _useCliff
```

Comment:
Actually, documenting this one could also help. It's also not clear to me at first what is it supposed to do.